run.c

```diff
@@ -221,16 +225,24 @@ void transformer(int token, int pos, Config* p, RunState* s, TransformerWeights*
     float* freq_cis_real_row = w->freq_cis_real + pos * head_size / 2;
     float* freq_cis_imag_row = w->freq_cis_imag + pos * head_size / 2;
 
+    //tmp files to read the current W being multiplied. Directly passing w->k .. doesn't work due to no support for virtual address offloading
```

Comment:
A thing to note with data is that moving it to and from GPU is an overhead. And we do it here in a loop by calling matmul in each iteration. And then in matmul we copy weights to the GPU. It would be efficient to try and map the whole weights to the GPU here before the loop by using `#pragma omp target data map` and then call matmul without having to copy data to GPU on each iteration. One can probably create a buffer for all weights, map it to GPU, then call matmul but I am not sure on this one. Just a direction worth exploring.

A reference here: [OpenMP CUDA](https://on-demand.gputechconf.com/gtc/2018/presentation/s8344-openmp-on-gpus-first-experiences-and-best-practices.pdf)

Reply:
I completely agree. It only makes sense if all memory is mapped on GPU memory

Reply:
Thanks for the suggestions, will take a look

Reply:
Made two changes

1. Offloading all weights to GPU at once. Removed all the `tmp` variables defined

```
#pragma omp target data map(to:weights.wq[0:size_qkvO],weights.wk[0:size_qkvO],weights.wv[0:size_qkvO],weights.wo[0:size_qkvO],weights.w1[0:size_ffn],weights.w3[0:size_ffn],weights.w2[0:size_ffn],weights.wcls[0:size_cls])
while (pos < steps) {

// forward the transformer to get logits for the next token
transformer(token, pos, &config, &state, &weights);
....
```


2. Using variant of matmul loop in [comment](https://github.com/karpathy/llama2.c/pull/166#discussion_r1278164971). New matmul is below. Run state variables are still copied to GPU. 
```python
void matmul(float* xout, float* x, float* w, int n, int d) {
    // W (d,n) @ x (n,) -> xout (d,)
    // by far the most amount of time is spent inside this little function
    int I;
    #pragma omp target data map(to:x[0:n]) map(from:xout[0:d])
    #pragma omp target teams distribute
    for (i = 0; i < d; i++) {
        float val = 0.0f;
        #pragma parallel for reduction(+:val) 
        for (int j = 0; j < n; j++) {
            val += w[i * n + j] * x[j];
        }
        xout[i] = val;
    }
}
```

Profiling. tok/s went from 20 to 80 currently. 
Next is try to allocate run state to GPU also & make the updates in it consistent with the value mapped to GPU.

Reply:
Wow this was quick again :). Thank you, this is really interesting to follow. This code at least deserves a fork IMHO. I am curious, what are your specs and which model did you profile?